### PR TITLE
fix Issue 14572 - cannot build dmd from source anymore: 'g++ -m64: No…

### PR DESCRIPTION
**Pull request by WalterBright**
… such file or directory'

Both gcc and clang support the "-x c++" switch.

https://issues.dlang.org/show_bug.cgi?id=14572

### CODE REVIEW

**Comment by WalterBright** [comment]
@yebblies this seems to break the DDMD build, I need some help with that, please

**Comment by yebblies** [comment]
The ddmd build command
`CC=$(HOST_CC) $(HOST_DC_RUN) $(MODEL_FLAG) $(DSRC) -ofddmd newdelete.o glue.a backend.a -vtls -J.. -d $(DFLAGS)`
needs to be changed so that `CC` is set to either `g++` or `gcc -x c++`, otherwise it won't link the c++ libs and then you get linker errors.

**Comment by WalterBright** [comment]
Thanks. Let's see if it works now...

**Comment by WalterBright** [comment]
That's not working, either. Can you please be more explicit about what change is needed for posix.mak to make ddmd build?

**Comment by yebblies** [comment]
It's expanding to this:
```
CC=gcc -x c++ /home/braddr/sandbox/at-client/release-build/install/freebsd/bin32/dmd -m32 access.d aggregate.d aliasthis.d apply.d argtypes.d arrayop.d arraytypes.d attrib.d builtin.d canthrow.d dcast.d dclass.d clone.d cond.d constfold.d cppmangle.d ctfeexpr.d declaration.d delegatize.d doc.d dsymbol.d denum.d expression.d func.d hdrgen.d id.d identifier.d imphint.d dimport.d dinifile.d inline.d init.d dinterpret.d json.d lexer.d link.d dmacro.d dmangle.d mars.d dmodule.d mtype.d opover.d optimize.d parse.d sapply.d dscope.d sideeffect.d statement.d staticassert.d dstruct.d target.d dtemplate.d traits.d dunittest.d utf.d dversion.d visitor.d lib.d nogc.d nspace.d errors.d tokens.d globals.d escape.d root/aav.d root/outbuffer.d root/stringtable.d root/file.d root/filename.d root/speller.d root/man.d root/response.d intrange.d complex.d entity.d backend.d root/array.d root/longdouble.d root/rootobject.d root/port.d root/rmem.d -ofddmd newdelete.o glue.a backend.a -vtls -J.. -d  -O -inline
-x: not found
```

Which is setting `CC` to `gcc`, not `gcc - c++`.  I assume it will work with some kind of quotes around it somewhere.

**Comment by WalterBright** [comment]
I don't understand what that line is trying to do. Set CC in the parents environment? Set CC in the makefile? Why would it not accept the -x switch, but accept the -m32 switch?

**Comment by yebblies** [comment]
Set CC for that run of DMD, so dmd links using the c++ compiler.  It's not accepting any -m32 switch.

File: src/posix.mak [hunk 1]
```diff
@@ -34,9 +34,9 @@ LDFLAGS=-lm -lstdc++ -lpthread
 #ifeq (osx,$(OS))
 #	HOST_CC=clang++
 #else
-	HOST_CC=g++
+	HOST_CC=gcc
 #endif
-CC=$(HOST_CC) $(MODEL_FLAG)
+CC=$(HOST_CC) -x c++ $(MODEL_FLAG)
```
Comment (JinShil) [inline src/posix.mak]: This will not work.  The offending problem is appending the `$(MODEL_FLAG)`
Proof:  Run this:

**Comment by JinShil** [comment]
I believe I have a solution at https://github.com/D-Programming-Language/dmd/pull/4772

**Comment by JinShil** [comment]
> The ddmd build command
CC=$(HOST_CC) $(HOST_DC_RUN) $(MODEL_FLAG) $(DSRC) -ofddmd newdelete.o glue.a backend.a -vtls -J.. -d $(DFLAGS)
needs to be changed so that CC is set to either g++ or gcc -x c++, otherwise it won't link the c++ libs and then you get linker errors.

That DDMD build command is doing two things at once:  Setting the CC environment variable, and running the host D compiler.  It confused the hell out of me, too.  You can do `CC='$(HOST_CC) -x c++' $(HOST_DC_RUN) ...` but then you get linker errors looking for things like `__cxa_guard_acquire`.  So it appears that `-x c++` only works for the compilation stage, not the linker stage.  If you add `-L-lstdc++`, then it will build and link.

**Comment by MartinNowak** [comment]
The obvious solution to fix mixing up C compilers and C++ compilers is to use the established CC and CXX variables.

**Comment by JinShil** [comment]
Well, its a little less obvious when your .c files contain c++ code.  Anyway, hopefully this will all go away soon with DDMD (2.069 I presume).

**Comment by JinShil** [comment]
This can be closed.

**Pull request closed without merging**